script/GameObjects/Elements/Components/Capacitor.lua

```diff
@@ -0,0 +1,145 @@
+
+  -- OK, back now to what Capacitor actually requires
+  self.name       = Config.gen.compCapacitorStats.name
+  self.healthCurr = Config.gen.compCapacitorStats.healthCurr
+  self.healthMax  = Config.gen.compCapacitorStats.healthMax
+  self.chargeCurr = Config.gen.compCapacitorStats.chargeCurr
+  self.chargeMax  = Config.gen.compCapacitorStats.chargeMax
+  self.chargeRate = Config.gen.compCapacitorStats.chargeRate
+--printf("Register: Capacitor name = '%s', type = %s, handler = %s", self.name, Event.Update, Capacitor.updateCapacitor)
+  self:register(Event.Update, Capacitor.updateCapacitor)
+end)
+
+function Capacitor:getSocketType ()
+  return SocketType.Capacitor
+end
+
+function Capacitor:getName ()
+  return self.name
+end
+
+function Capacitor:setName (newName)
+  self.name = newName
+end
+
+function Capacitor:damageHealth (amount)
+  if self.healthCurr - amount < 1e-6 then
+    self.healthCurr = 0.0
+  else
+    self.healthCurr = self.healthCurr - amount
+  end
+--printf("Vessel %s capacitor takes %s damage, %s remaining", self:getName(), amount, self.healthCurr)
+
+  -- Reduce maximum possible charge due to damage
+  -- We could also reduce the charge rate, but let's be nice for now
+  local maxCharge = self.chargeMax * (self.health/self.healthMax)
+  if self.chargeCurr > maxCharge then self.chargeCurr = maxCharge end
+  Capacitor:setCharge(self.chargeCurr, maxCharge, self.chargeRate)
+end
+
+function Capacitor:getHealth ()
+  return self.healthCurr or 0.0
+end
+
+function Capacitor:getHealthMax ()
+  return self.healthMax or 0.0
+end
+
+function Capacitor:getHealthPercent ()
+  if self.healthMax < 1e-6 then return 0.0 end
+  return 100.0 * self.healthCurr / self.healthMax
+end
+
+function Capacitor:setHealth (value, max)
+  self.healthCurr = value
+  self.healthMax = floor(max)
+end
+
+function Capacitor:discharge (value)
+  local undischargedAmount = 0
+  local chargeRemaining = self.chargeCurr
+  local newCharge = chargeRemaining - value
+
+  if newCharge < 0 then
+    undischargedAmount = 0 - newCharge
+    newCharge = 0.0
+  else
+    undischargedAmount = 0
+  end
+
+  self.chargeCurr = newCharge
+--printf("Entity %s discharges %s, %s charge remaining, %s undischarged",
+--self:getName(), value, self.chargeCurr, undischargedAmount)
+  return undischargedAmount
+end
+
+function Capacitor:getCharge ()
+  return self.chargeCurr or 0
+end
+
+function Capacitor:getChargeMax ()
+  return self.chargeMax or 0
+end
+
+function Capacitor:getChargeRate ()
+  return self.chargeRate or 0
+end
+
+function Capacitor:getChargePercent ()
+  return 100.0 * self.chargeCurr / self.chargeMax
+end
+
+function Capacitor:setCharge (value, max, rate)
+  self.chargeCurr = value
+  self.chargeMax = max
+  self.chargeRate = rate
+end
+
+function Capacitor:updateCapacitor (state)
+  if not self:getParent():isDestroyed() then
+    if self.chargeCurr < self.chargeMax then
+      local timeScale = 1.0
+      if GameState.paused then
+        timeScale = 0.0
```

Comment:
might be good to establish GameState.currentTimeScale and use it here / globally -> Application.lua

Reply:
I'll look at this.

Reply:
I think eventually we may wind up removing testing services like time acceleration and instantJobs. This means that GameState.debug.timeAccelFactor, used here and elsewhere, may wind up simply being removed in the future when we're close to releasing LTR as an actual game playable by people not following it in our Discord server. Until then, however, it's really useful for testing! I'd like to leave this intact for now.

Reply:
Sounds good. The currentTimeScale state might still be useful if nonetheless. :)